lib/Beam/Emitter.pm

```diff
@@ -15,20 +15,27 @@ use Beam::Listener;
 # The event listeners on this object, a hashref of arrayrefs of
 # EVENT_NAME => [ Beam::Listener object, ... ]
 
-=attr listeners
-
-An array containing the listeners which have subscribed to this
-emitter.  The array elements are instances of classes which inherit
-from L<Beam::Listener>
-
-=cut
-
-has listeners => (
+has _listeners => (
     is      => 'ro',
     isa     => HashRef,
     default => sub { {} },
 );
 
+=method listeners( event_name )
+
+Returns an array containing the listeners which have subscribed to the
+specified event from this emitter.  The array elements are instances
+of sub-classes of L<Beam::Listener>.
+
+=cut
+
+sub listeners {
+
+    my ( $self, $name ) = @_;
+
+    return $self->_listeners->{$name} || [];
```

Comment:
Yeah, that's a possible problem, but I don't think usage of this will reach that point. We're likely talking many thousands of items before that becomes a true bottleneck (though now I am curious to run some microbenchmarks to see...). Besides, making a shallow copy like that does the same thing, really: all the things in `$arrayref` go on the stack as it's dereferenced in list context, then the stack gets used to create the arrayref. I imagine `perl` makes some optimizations in both situations, and until I need otherwise, I generally trust `perl` to do it as fast as it thinks it can. So it's a tradeoff between potential bugs or potential performance problems, and at the moment I think those potential bugs are a bit more potential ;)

Were this something like an ETL or something that we could very obviously pass off ownership of the arrayref and had good reason to believe that the data set would be extremely large, I'd likely fall the other way. Or, if Perl had built-in and fast immutable iterable like Python's tuple...